### DISCUSSION TITLE
[FEATURE REQUEST] LaTeX Math in Markdown

### DISCUSSION THREAD

**Discussion by Polda18** [community/community#23083] [New to GitHub]
<p>Hi. I am using a math snippet in my markdown readme file in my repository. This is the code:</p>
<pre><code>(...)
  * Experiences and leveling =&gt; every entertainment bot shall have this feature and this is
    no exception! The formula for leveling taken from experiences may change in future.
    Currently, no formula is used. I am, however, thinking of this one:
    $$n = x^{l + 1} - x^l$$
    $n$ is the required experience value to gain, $l$ is current level value, $x$ a further
    unspecified constant (it's value is not yet determined). The cooldown for experience gain,
    leveling and command fetch may be 15 seconds by default.
(...)
</code></pre>
<p>This is how it looks like in my markdown preview as displayed by VS Code Markdown and Markdown MathJax plugins: </p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/d113cceb632d34a47f910f1213ba34f963140f72.jpeg" data-download-href="/uploads/short-url/tPAdfbzxveA8NgLKoKSTDTIHIJk.jpg?dl=1" title="github_markdown_latex_showcase1.jpg" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/25118993/181074726-327ec0a6-c59a-4b8c-86cc-06d093a1e59a.jpg" alt="github_markdown_latex_showcase1.jpg" data-base62-sha1="tPAdfbzxveA8NgLKoKSTDTIHIJk" width="690" height="388" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/d113cceb632d34a47f910f1213ba34f963140f72_2_690x388.jpg, https://aws1.discourse-cdn.com/github/optimized/1X/d113cceb632d34a47f910f1213ba34f963140f72_2_1035x582.jpg 1.5x, https://aws1.discourse-cdn.com/github/optimized/1X/d113cceb632d34a47f910f1213ba34f963140f72_2_1380x776.jpg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/d113cceb632d34a47f910f1213ba34f963140f72_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">github_markdown_latex_showcase1.jpg</span><span class="informations">1920×1080 477 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>But this is how it looks like on GitHub: </p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/6b9e117a311c49660e50defbd82e9531a4a23948.jpeg" data-download-href="/uploads/short-url/fm1KrkH279qVEOiQV9rukVQovTq.jpg?dl=1" title="github_markdown_latex_showcase2.jpg" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/25118993/181074727-751962ef-1af0-4483-a4b4-d74e6d3289cc.jpg" alt="github_markdown_latex_showcase2.jpg" data-base62-sha1="fm1KrkH279qVEOiQV9rukVQovTq" width="690" height="388" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/6b9e117a311c49660e50defbd82e9531a4a23948_2_690x388.jpg, https://aws1.discourse-cdn.com/github/optimized/1X/6b9e117a311c49660e50defbd82e9531a4a23948_2_1035x582.jpg 1.5x, https://aws1.discourse-cdn.com/github/optimized/1X/6b9e117a311c49660e50defbd82e9531a4a23948_2_1380x776.jpg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/6b9e117a311c49660e50defbd82e9531a4a23948_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">github_markdown_latex_showcase2.jpg</span><span class="informations">1920×1080 516 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>GitHub does not render LaTeX Math notations. Why? As far as I know, it did in past. Why was that feature removed? Could you bring it back again? It breaks the page intended use. Users can get it working in GitHub pages with few extra lines of code in _ <strong>_config.yml</strong> _, but unfortunately that doesn’t work with the internal markdown renderer at the repo view. I could simply use a Chrome extension, but the downfall of the extension is that it tries to interpret the LaTeX code even inside code blocks or in in-line code notations, which is wrong.</p>

**Comment by BelHFL** [comment]
<p>Hi <a class="mention" href="/polda18">@polda18</a> ,</p>
<p>Thanks for this feedback! We’re always working to improve GitHub and the GitHub Community Forum, and we consider every suggestion we receive. I’ve logged your feature request in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>
<p>Cheers!</p>

**Comment by Polda18** [comment]
<p>Thank you :blush:</p>
<p>It’s really bad that math does not get recognised in inner GitHub markdown renderer. If I want to put math equation to the readme, it doesn’t display correctly (as intended), it displays as is written in code.</p>
<p>Thank you for considering to include it :+1:</p>

**Comment by ghost** [comment]
<p>Thank you for bringing attention to this topic to the appropriate teams! One thing I suggest is using <a href="https://katex.org/" rel="nofollow noopener">KaTeX</a> instead of MathJax, if math support were to be implemented in the near future, due to its efficiency compared to MathJax. Although <a href="https://katex.org/docs/supported.html" rel="nofollow noopener">KaTeX</a> does not support all mathematical functions MathJax supports, it does heavily reduce the loading time for markdown files infused with mathematical operations!</p>
<p>Best regards!</p>

**Comment by roel1975** [comment]
<p>Thanks, you’re the first one who reacts positively</p>

**Comment by roel1975** [comment]
<p>Thanks for the advice, but this one I give away. I didn’t use either of the two</p>

**Comment by roel1975** [comment]
<p>I think I know why. He has to hide so massively much lets say data, that it will show up real soon. I am speaking of experience now for what he did to me for putting me on that cc list. But I have already contacs with the source.</p>

**Comment by yxs** [comment]
<p>Is it possible for GitHub to switch the rendering engine to <a href="https://github.com/yuin/goldmark" rel="nofollow noopener">goldmark</a> ? It has great math extension <a href="https://github.com/litao91/goldmark-mathjax" rel="nofollow noopener">goldmark-mathjax</a> </p>

**Comment by michaelmyc** [comment]
<p>I realize that some of the reasons why this wasn’t implemented earlier are because of the speed of MathJax and the maintenance issues that come with it. I think Katex may help with the speed issue, and even if it still isn’t fast enough, GitHub can detect files with a lot of math equations and have the user toggle rendering on if needed while directly rendering files that have fewer equations. I’m not sure what sorts of maintenance issues it may bring, but since jupyter notebooks on GitHub already renders equations, it should not be that difficult to extend that to all of GitHub. </p>

**Comment by zhangddjs** [comment]
<p>I’m looking forward to this feature been implemented. I am working on a markdown book with a good num of latex style characters, the process went 20% while I found that latex not shown on github, it is hard for me to repair this issue, which confused me so much.</p>

**Comment by Makeshiftdev-c** [comment]
<p>It’s a bit late but hopefully this helps <a href="https://md-math.netlify.app/" rel="noopener nofollow ugc">https://md-math.netlify.app/</a> . I’m progressively enhancing it and slowly refactoring, so if anyone has any issues with it, drop me a line and I’ll mediate as soon as.</p>

**Comment by realliyifei** [comment]
<p>I don’t understand. Is it really technically hard to add the feature? Given there’s already extension in chrome, I don’t think it’s hard. And this feature is obviously important to lots of GitHub users, specially those who in DS/ML domains.</p>

**Comment by dholstius** [comment]
<p>+1 for this request. It’d be really nice for a slew of users who use GitHub repos for technical work of all kinds.</p>

**Comment by jeremy-rifkin** [comment]
<p>+1<br>
I’ve written a <a href="https://github.com/jeremy-rifkin/markdown-math-gh-compiler" rel="noopener nofollow ugc">tool</a> to compiler markdown with latex into github friendly markdown using the a latex image renderer. But, because the image height is inconsistent and the rendering is not done locally the equations are often not vertically aligned nicely, e.g.: <a href="https://gist.github.com/a-rodin/fef3f543412d6e1ec5b6cf55bf197d7b#gistcomment-3793409" rel="noopener nofollow ugc">https://gist.github.com/a-rodin/fef3f543412d6e1ec5b6cf55bf197d7b#gistcomment-3793409</a>.</p>
<p>It’d be super nice to have github support latex equations natively, though for now the hacks such as my compiler or PurplePi will do.</p>

**Comment by jeremy-rifkin** [comment]
<p>I forgot to include this the other day: Aside from alignment issues the big pitfall with using the renderer is that it is not theme-aware and will be unreadable on dark theme.</p>

**Comment by nschloe** [comment]
<p>I created a small Chrome extension, <a href="https://github.com/nschloe/xhub" rel="noopener nofollow ugc">xhub</a>, to add math (and more) to GitHub pages.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/f/0/f04776ff5a1e86806190eef9349002358488c198.png" data-download-href="/uploads/short-url/yhBHKmbu0S0wqTfSgzK7qOuKYmA.png?dl=1" title="screenshot" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/181628/181074735-78959a4f-a507-4f1b-8f16-3aec73761c21.png" alt="screenshot" data-base62-sha1="yhBHKmbu0S0wqTfSgzK7qOuKYmA" width="690" height="257" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/f/0/f04776ff5a1e86806190eef9349002358488c198_2_690x257.png, https://aws1.discourse-cdn.com/github/optimized/3X/f/0/f04776ff5a1e86806190eef9349002358488c198_2_1035x385.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/f/0/f04776ff5a1e86806190eef9349002358488c198.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/f/0/f04776ff5a1e86806190eef9349002358488c198_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">screenshot</span><span class="informations">1234×461 23.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>Pros:</p>
<ul>
<li>You don’t have to set up anything in your repo, just use math in your Markdown (sytax from <a href="https://docs.gitlab.com/ee/user/markdown.html#math" rel="noopener nofollow ugc">GitLab</a>):<pre><code class="lang-markdown">Some display math:
```math
e^{i\pi} + 1 = 0
```
and some inline math, $`a^2 + b^2 = c^2`$.
</code></pre>
</li>
<li>It works on light and dark backgrounds alike.</li>
<li>You can copy-and-paste the math</li>
</ul>
<p>Cons:</p>
<ul>
<li>You have to install a browser extension once.</li>
</ul>

**Comment by I-use-IT** [comment]
<p>Very nice solution <a class="mention" href="/nschloe">@nschloe</a>!<br>
But still +1 for native GitHub support.</p>

**Comment by cxcorp** [comment, marked as answer]
<p>FYI they just released LaTeX support (using MathJax) yesterday: <a href="https://github.blog/changelog/2022-05-19-render-mathematical-expressions-in-markdown/" class="inline-onebox" rel="noopener nofollow ugc">Render mathematical expressions in Markdown | GitHub Changelog</a></p>

**Comment by ettaboyle** [comment]
<aside class="quote no-group" data-username="cxcorp" data-post="19" data-topic="14214">
<div class="title">
<div class="quote-controls"></div>
 cxcorp:</div>
<blockquote>
<p>Render mathematical expressions in Markdown | GitHub Changelog</p>
</blockquote>
</aside>
<p>:clap:t2: Welcome to the forums <a class="mention" href="/cxcorp">@cxcorp</a> and thank you for sharing that solution!! Appreciate you closing the loop for folks :sparkles:</p>

**Comment by Polda18** [comment]
<p>Thanks, finally! I can then uninstall the Chrome extension :slight_smile:</p>
<p>EDIT: When is this feature going to roll out though? I uninstalled the MathJax for GH Chrome extension already, and the math expressions still don’t work, no beta feature for it is available yet.</p>
